### ISSUE TITLE
Startup Flow and Chaining

### ISSUE THREAD

**Issue by alvyxaz** [alvyxaz/barebones-masterserver#21]
Hey! 

I want to start a discussion to see if any of you have any tips or thoughts on how we could better manage the startup of the framework.

While working with the framework, I've noticed that a big part of the "difficulty" is understanding the flow of events and how things startup. This is getting slightly worse with each update, as the number of things to "initialize" increases.

Currently, the flow is managed with a number conditional statements, and unity's built-in methods (Start / Awake). The problem is that components get initialized in a random order (Awake() and Start() methods), and there's nothing "binding" them together - each one has to take care of itself, and see if other components are initialized, skipping frames and etc. This problem becomes very apparent when you try to implement, for example, logging wrapper - it should be initialized first, but where should I do that?

What I want to propose is having a very simple "chaining" mechanism, and removing the startup from components themselves.

What it would look like? There would be some sort of a **starter** component (your custom script) in the scene, which would initiate a chain of events sequentially. What it would look like?

``` C#
void Awake() {
	
	// Master Server
	var masterChain = new BmChain();
	chain.Do((done) => {
		// 1. Initialize logs
		Logs.Initialize();
		done();
	})
	.WhenDone((done) => {
		// 2. Start a master server
		// 3. Initialize master components
		done();
	});
	
	// Spawner Server
	var spawnerChain = new BmChain();
	chain.Do((done) => {
		// 1. Initialize logs
		Logs.Initialize();
		done();
	})
	.WhenDone((done) => {
		// 2. Check if executable is in the right place
		done();
	})
	.WhenDone((done) => {
		// 3. Connect to master server
	})
	.WhenDone((done) => {
		// 4. Register to master server
	});
	
}
```

It might not look like much, but it makes the whole startup a lot more clear and easier to understand. This functionality could be extended to support "**merging**" chains. This would allow you to write your own chains, and merge when appropriately (for example, when starting both **Master server** and **Spawner**, you could simply merge the two chains together, or use them separately when needed).

Here's what a game debugging flow would look like:

``` C#
void DebugGameServerInEditor() {
	var masterChain =...
	var spawnerChain =...
	var gameServerStartChain=...
	
	var debugChain = new BmChain();
	
	debugChain.Do(Logs.Initialize)
	.WhenDone(masterChain) 	// Start master
	.WhenDone(spawnerChain)	// Start spawner (just for example)
	.WhenDone(gameServerStartChain) // Start a game server
	.WhenDone((done) => {
		// 1. Connect as a client to master server
	})
	.WhenDone((done) => {
		// 2. Log in (authenticate)
	})
	.WhenDone((done) => {
		// 3. Join game server
	});
	
	// Execute the chain
	debugChain.Execute();
}
```

With the chaining approach, there would be no more "initialization magic" (when god knows which components listen to which events) - everything flows just as described in the chain, and you don't have to dig through numerous files to see what starts first, and what follows after that

This is just a basic idea, the exact implementation would probably be a bit different, but **I'd really like to hear what you think**.

Cheers!

**Comment by bigcarlito** [comment]
Personally, I like the current setup where each (master/spawner/game) server object is somewhat autonomous and uses it's own logic to determine if it should start up or not.  I feel like it's simpler and easier to understand.  I'm always a proponent of using the least code to accomplish the task.  Is there a reason you do not utilize the 'script execution order' feature of Unity for a reliable order of initalization?

P.S.  Thanks for your work on this, I am just getting into this and it is working well for me so far, so expect more contributions (and questions) from me in the future :+1:

**Comment by cdngater** [comment]
I do like the idea of BmChain, but maybe not for what your thinking.  I am tending to agree with bigcarlito, each server object should be autonomous.  In a previous incarnation of my own code, I had the same need.  What I ended up doing was creating module coroutines that did the initialising.   That way, they could wait for the status of a dependent module to be ready if they needed to without holding up the initialization of other modules.  Also created a global ModuleManager object so the modules could find each other.

Pseudo Code

```

LoggingModule:Start() {
   StartCoroutine(_init);
}

IEnumerator LoggingModule::_init()
{
   state = booting;
   
   state = initialized;
}


OtherModule::Start() {
  StartCoroutine(_init);
}

IEnumerator OtherModule::_init()
{
   state = booting;
   logModule = GetModule<LoggingModule>();
   while (logModule.state != initialized)
      yield return null;
   state = initialized;
}

```

**Comment by 0F00** [comment]
```
private void Start
{
Barebones.StartMasterServer(MasterServerConfig, MSCallback_SUCCESS, MSCallback_FAILED); 
}
private override MSCallback_SUCCESS("some other stuff u think is important", MasterServer ms)
{
Barebones.SpawnerChain("some stuff etc u think is important", SCCallback_SUCCESS(ms), SCCallback_FAILED(ms));
}
private override MSCallback_FAILED(MasterServerException ex)
{
Debug.Log(ex);
}
```

I Think so is the easiest way and everyone understand what did this mean.
alvyxaz what do and the other think about this?

**Comment by cdngater** [comment]
0F00, that does not really solve the issue alvyxaz is trying to solve.  If you have a Module that requires two other modules to be up and running before it can initialize, how would you do it with your code?  I think this would complicate things a lot more.  

Something like this is what your suggesting
```
private void Start()
{
	Barebones.StartLogServer(LogServerConfig, LS_SUCCESS, LS_FAILED);
}

private void LS_SUCCESS(LogServer ls)
{
	BareBones.StartMasterServer(MasterServerConfig, MS_SUCCESS, MS_FAILED);
}

private void LS_FAILED(Exception e)
{
	Debug.LogError(e);
}

private void MS_SUCCESS(MasterSever ms)
{
	BareBones.StartAuthenticationServer(AuthenticationServer, AS_SUCCESS, AS_FAILED);
}

private void MS_FAILED(Exception e)
{
	Debug.LogError(e);
}

private void AS_SUCCESS(AuthenticationServer as)
{
	'' Do something here
}

private void AS_FAILED(Exception e)
{
	Debug.LogError(e);
}
'''

One major issue I see is having to create Start* functions in BareBones for each thing thats required to start.
Now lets change it up some to indicate some dependancy, see code below for the QUESTION.

'''
private void Start()
{
	Barebones.StartDBServer(DBServerConfig, DB_SUCCESS, DB_FAILED);
	BareBones.StartMasterServer(MasterServerConfig, MS_SUCCESS, MS_FAILED);
}

private void DS_SUCCESS(DBServer db)
{
	
}

private void DS_FAILED(Exception e)
{
	Debug.LogError(e);
}

private void MS_SUCCESS(MasterSever ms)
{
	**// How do I start the AuthenticationServer thats dependent on DatabaseServer and MasterServer being already started?  But Remember, DatabaseServer and MasterServer don't care about each other.**
	// BareBones.StartAuthenticationServer(AuthenticationServer, AS_SUCCESS, AS_FAILED);
}

private void MS_FAILED(Exception e)
{
	Debug.LogError(e);
}

private void AS_SUCCESS(AuthenticationServer as)
{
	'' Do something here
}

private void AS_FAILED(Exception e)
{
	Debug.LogError(e);
}
'''

**Comment by 0F00** [comment]
when u combine directly in the framework them that u get only 1 access and more expections it work?

**Comment by alvyxaz** [comment]
Hey guys!

After careful considerations, I think, for now, we can leave it as it is :). 

In the future, if this "startup" becomes more difficult to handle, we might come up with something better.

Thanks for sharing your opinions! I really appreciate it ;)

**Comment by jeff-odg** [comment]
Something that might help is using events for when different components start up. So when the MasterServer is up send a masterServerUp event. Same for database connections, the connection modules etc.  

Once those are available you can tie it into the Unity Events and trigger off of them.  

Just an idea.

J.

**Comment by alvyxaz** [comment]
Right before publishing framework to the asset store, I had an idea to make a bunch of events, to which certain modules could subscribe and etc. This was mainly to achieve decoupling.

What I ended up with was a huge number of events, and understanding how everything works and ties together became pretty much impossible.

The thing that I've learned from all of this was - don't use events to control "the flow" (of startup, bootrapping, or anything that requires some kind of ordering), because when things get a bit more complex, they become very hard to follow (for example, instead of following code normally, through invocations, you end up having to track all of the events, see what's listening to them, what's "firing" and what's not, which gets a bit tedious)

However, I think it's a good idea to have more events just to notify about state changes of various components (as you've mentioned, master server "up" event could be quite handy). 

And using Unity Events is a good idea too ;)
